### PR TITLE
Fixes Header._keyword_from_index

### PR DESCRIPTION
**Pull request by mohanagr**
This addresses issue #5827

### CODE REVIEW

**Comment by pllim** [comment]
@mohanagr , thanks for tackling this. This needs a test and a change log. Please see http://docs.astropy.org/en/latest/development/workflow/development_workflow.html

File: astropy/io/fits/header.py [hunk 1]
```diff
@@ -1694,11 +1694,11 @@ def _keyword_from_index(self, idx):
         """
 
         if idx < 0:
-            idx += len(self._cards) - 1
+            idx += len(self._cards)
 
         keyword = self._cards[idx].keyword
         keyword = Card.normalize_keyword(keyword)
-        repeat = self._keyword_indices[keyword].index(idx)
+        repeat = len(self._keyword_indices[keyword]) - 1
```
Comment (pllim) [inline astropy/io/fits/header.py]: Why did you change this?

Comment (mohanagr) [inline astropy/io/fits/header.py]: header._keyword_indices[keyword] gives a list of the indices where a particular keyword is present. Now it did not give the length as explained in #5827 . 

```>>> hdr = pf.Header([ ('MOONFRAC', 0.604829), ('MOONDIST', 82.5171118)])
>>> hdr.append()
>>> hdr.append()
>>> hdr._keyword_indices['']
[2,3]
```
So number of times a header is repeated is just `len(hdr._keyword_indices[''])-1` which in this case for '' will be 1 and for others will be 0.

File: astropy/io/fits/header.py [hunk 2]
```diff
@@ -1694,11 +1694,11 @@ def _keyword_from_index(self, idx):
         """
 
         if idx < 0:
-            idx += len(self._cards) - 1
+            idx += len(self._cards)
```
Comment (saimn) [inline astropy/io/fits/header.py]: This change seems right, though it's bizarre that nobody noticed this before as it is there since a long time (it comes from a Pyfits 3.2 merge, 3 years ago!). So this clearly needs a test ;-)

File: astropy/io/fits/header.py (hunk 1 above)
Comment (saimn) [inline astropy/io/fits/header.py]: I don't think this needs to be changed: though the docstring is unclear about `repeat`, from the usage of it in the `compressed` module it seems to be the sub-index of the keyword in the list of repeated keyword. At least it needs to be unique, which explains why with your change some tests are failing.

Comment (mohanagr) [inline astropy/io/fits/header.py]: Exactly, I made the change based on docstring and what @rsiverd said in the issue. Specifically 
```
...
or most keywords the repeat will always be zero, but it may be greater than zero for keywords that are duplicated (especially commentary keywords).
```
What do you suggest I do? I am seeing the compressed module. Was unaware of that.

Comment (saimn) [inline astropy/io/fits/header.py]: In my understanding, the "incorrect repeat count" mentioned in the original issue is fixed with the first line of your patch (the `idx` value). So there is no need to modify the repeat count itself.

Comment (mohanagr) [inline astropy/io/fits/header.py]: So I revert this change and add test for the function?

File: CHANGES.rst [hunk 3]
```diff
@@ -141,6 +142,8 @@ Bug Fixes
 
 - ``astropy.io.fits``
 
+  - Fixed ``fits.Header._keyword_from_index()`` to give correct keyword and repeat number [#5827]
```
Comment (pllim) [inline CHANGES.rst]: 1. Change log for bug fix should fall under 1.3.1. 
2. Include this PR number (not just the original issue). For example, `[#5827, #5852]`. Also does not hurt to add a period at the end of "number".

**Comment by pllim** [comment]
Your test violated some sort of FITS standard, thus the failing tests. Please fix.

**Comment by saimn** [comment]
The failing tests (`TestCompressedImage.test_compression_header_insert` and `TestCompressedImage.test_compression_header_set_before_after`) are related to the change with the `len(self._cards) - 1`, not to the test that you have added (which tests only the private method `_keyword_from_index` but it doesn't test the actual use of this method).

So obviously changing the behavior of `_keyword_from_index` breaks the compressed image code. Things are never simple with FITS 😆 
For reference, a failing build : https://travis-ci.org/astropy/astropy/jobs/206741921#L2458

**Comment by saimn** [comment]
Also it can be a great idea to run tests locally, to avoid spawning many builds on Travis. With your last commit it will fail again.

**Comment by mohanagr** [comment]
@saimn sorry about that bit of carelessness. Should the test be modified or should I try and fix what modifying the function is breaking?

**Comment by saimn** [comment]
@mohanagr - The second option seems more reasonable ! If you want to try it's up to you, but from the quick look I had it seems not so easy.

**Comment by rsiverd** [comment]
Thanks for looking into this. I would have liked to contribute a pull request directly but I'm currently traveling out of the country. Given the way that those routines are used throughout the ImageHeader and CompImageHeader classes, I suspected this might be a complicated fix.

**Comment by bsipocz** [comment]
@rsiverd - You're right fixing io.fits is generally always end up being complicated, and I think the only reason it got the ``package-novice`` tag, is that you've already did the heavy lifting by debugging and identifying the problematic line. Thank you for it.
If you are interested in contributing when you're back, please have a look around, there are plenty of issues to fix.

**Comment by mohanagr** [comment]
I am going through and understanding the part which was failing. Although I don't know what else will fail if I try and fix that.

**Comment by pllim** [comment]
Ops, should I remove the `novice` tag? I thought just fixing that line would be enough but apparently not...

**Comment by saimn** [comment]
@pllim - Yes you can remove this tag ;-)

I have a patch ready, just didn't have the time to push it sooner. @mohanagr - Do you want to continue to work on this ? As you like, just tell me ;-)

**Comment by mohanagr** [comment]
@saimn I'd like to give this a try.

**Comment by mohanagr** [comment]
@saimn I believe the bug lies :[https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/compressed.py#L172](bug)
The index being passed here is the one gotten for `self` not `self._table_header`. `_table_header` is a separate `Header` instance. I checked and saw that `self._keyword_indices` and `self._table_header._keyword_indices` are different and hence was getting an `IndexError`.

if I run `_, index = self._table_header._keyword_from_index(key)` there, the failing test passes. Although this makes some other tests fail I believe that is because I need to implement the same checks for `key` there too before getting index again, like [https://github.com/astropy/astropy/blob/master/astropy/io/fits/hdu/compressed.py#L156](check) (if key is int or tuple).

**Comment by mohanagr** [comment]
The tests have passed successfully locally.

**Comment by saimn** [comment]
@mohanagr - It works indeed, but there is another underlying issue. I don't know much about this module but here is my understanding:
As you noticed, CompImageHeader uses two header, a "compressed" one and a "normal" one, and tries to keep them synchronized. But things goes wrong in the insert and append methods, when dealing with blank cards. While adding a new card, a blank one must deleted in each header, but due to the inheritance between CompImageHeader and Header it is deleting two blank cards in the compressed header.
So here you are fixing the second deletion, but still two blanks card are deleted in the compressed header instead of one (which you can verify by calling `Header._countblanks` in a loop).

**Comment by mohanagr** [comment]
This should be closed now?

**Comment by mohanagr** [comment]
This should be closed now?

**Comment by saimn** [comment]
@mohanagr - Sorry, I wanted to comment here after submitted the PR but I forgot. So, I submitted the PR because it's easier to discuss the correct fix when the code is available. As I explained I *think* your fix is not correct as it does not preserve blank cards, but I'm not a `io.fits` maintainer so I will leave the decision to someone else ;-).
Anyway at this point having another pair of eyes would be useful.

**Comment by mohanagr** [comment]
@saimn Yes, I went through the `_useblanks` thing after your previous comment which I hadn't done before and think #5866 is the complete fix even though mine makes it works somehow.

**Comment by pllim** [comment]
Superseded by #5866. Thanks anyway. Your effort is appreciated.
